Fix/50850 Global Style applied correctly to the table block

faisal-alvi opened the pull request:
Closes #50850

## What?

This PR fixes the issue between the global styles and the block editor styles for the table block. Also refactors the `block.json` to use [the new Selectors API](https://make.wordpress.org/core/2023/07/17/introducing-the-block-selectors-api/).

## Why?

The global styles were only applied to the table tag but a few styles like spacing should be applied to the td tags.

## How?

This PR fixes the issue by updating the global styles CSS and the block editor-level styles and also by using the new Selectors API.

## Testing Instructions

1. Go to the site editor.
2. Add a table block with some contents
3. Go to Global Styles and change the table block border globally (color and width.)
4. See the global style changes are applied correctly.
5. Also confirm the global styles are overwritten at block settings level.

## Screenshots or screencast <!-- if applicable -->

![image](https://github.com/WordPress/gutenberg/assets/25176325/33289ae5-fa7e-4576-b77b-a5b8e328057e)

faisal-alvi committed:
Global style applied to table `td` border

faisal-alvi committed:
Merge branch 'trunk' into fix/50850

faisal-alvi commented:
@t-hamano Is this PR helpful? Can you please check and let me know if any changes are required?

t-hamano commented:
Sorry for the late feedback 🙇

To be honest, this problem is very complex.

This is because we need to be able to override styles set in the global styles at the block level. For this PR, please try the following scenario.

- Apply borders, font size, and background color to the Table block in the global styles.
- Insert a table block into the block editor.
- Apply borders, font size, and background color to the Table block.
- Probably no styles other than the border will be applied.

To ideally solve a problem like this, I think we should refactor with [the new Selectors API](https://make.wordpress.org/core/2023/07/17/introducing-the-block-selectors-api/). Additionally, we may also need to adjust [this CSS](https://github.com/WordPress/gutenberg/blob/1c7f3a8f97dcc47b6c1b61f633ccb0cf362b8268/packages/block-library/src/table/style.scss#L115-L198) regarding the border.

faisal-alvi committed:
update the logic to cover the rest of the areas

faisal-alvi commented:
@t-hamano thanks for the feedback, I see the areas of improvement and fixed them so that it should solve the issue. Can you please have a look?

However, one point to share here is that the padding from the Global styles is applied to the individual `<td>`s but at the block editor it is applied to the `<figure>` tag, this is another issue (not sure if someone has already reported but) this should be solved separately.

faisal-alvi commented:
I realized I did not use [the new Selectors API](https://make.wordpress.org/core/2023/07/17/introducing-the-block-selectors-api/), I'll try to refactor using it and update the PR soon.

faisal-alvi committed:
Merge remote-tracking branch 'upstream/trunk' into fix/50850

faisal-alvi committed:
use the new Selectors API and refactor table block.json

faisal-alvi committed:
a couple of minor changes

faisal-alvi commented:
@t-hamano I've updated the PR and refactored the table `block.json` to use [the new Selectors API](https://make.wordpress.org/core/2023/07/17/introducing-the-block-selectors-api/). Can you please have a look or assign someone else as a reviewer?
